### PR TITLE
Add the grid

### PR DESCRIPTION
**Pull request by thewilkybarkid**

### CODE REVIEW

File: vendor-extra/ContentPageBundle/src/Controller/ContentController.php [hunk 1]
```diff
@@ -62,7 +64,10 @@ function (ResponseInterface $response) use ($request) : Response {
                     return new Response(
                         $this->twig->render(
                             $this->template,
-                            $event->getContext() + ['title' => $event->getTitle(), 'content' => $event->getContent()]
+                            $event->getContext() + [
+                                'title' => $event->getTitle(),
+                                'content' => json_decode(json_encode($event->getContent()), true),
```
Comment (thewilkybarkid) [inline vendor-extra/ContentPageBundle/src/Controller/ContentController.php]: The problem is that Twig's `with` _only_ works with arrays, not `ArrayAccess` objects. There's probably a far better way round it.

Comment (stephenwf) [reply vendor-extra/ContentPageBundle/src/Controller/ContentController.php]: Is this the problem line: https://github.com/twigphp/Twig/blob/2.x/src/Node/WithNode.php#L43

Could be replaced with `is_iterable` if twig targeted only PHP 7+ not sure if theres a simple fix

Comment (thewilkybarkid) [reply vendor-extra/ContentPageBundle/src/Controller/ContentController.php]: Opened https://github.com/twigphp/Twig/pull/2901.

File: vendor-extra/ViewsBundle/src/Views/View.php [hunk 2]
```diff
@@ -5,17 +5,21 @@
...
+use function is_string;
 
 final class View implements JsonSerializable
 {
     private $arguments;
+    private $context;
     private $template;
 
-    public function __construct(?string $template, array $arguments = [])
+    public function __construct(?string $template, array $arguments = [], array $context = [])
```
Comment (thewilkybarkid) [inline vendor-extra/ViewsBundle/src/Views/View.php]: Will extract this change into a separate PR.

File: vendor-extra/ContentPageBundle/src/Controller/ContentController.php (hunk 1 above)
Comment (thewilkybarkid) [reply vendor-extra/ContentPageBundle/src/Controller/ContentController.php]: Merged in, but only for `Traversable`. Will see if there's a way round it until that's released.

Comment (thewilkybarkid) [reply vendor-extra/ContentPageBundle/src/Controller/ContentController.php]: Patching in 2e470a8853c8bde1cd9eef7fc3499b47b30dcf93.

File: composer.json [hunk 3]
```diff
@@ -74,6 +75,12 @@
         "sort-packages": true
     },
     "extra": {
+        "composer-exit-on-patch-failure": true,
+        "patches": {
+            "twig/twig": {
+                "Allow Traversable objects to be used in the with tag": "https://github.com/twigphp/Twig/commit/650bc7c262d0c84ece3c472c9302bf416475f72b.patch"
```
Comment (thewilkybarkid) [inline composer.json]: Not ideal as this isn't the merged change, will update when it appears in the Twig `2.x` branch. (Then remove entirely when it's released.)

File: vendor-extra/JatsContentBundle/src/EventListener/TitleListener.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+final class TitleListener
+{
+    public function onCreatePage(CreateContentPageEvent $event) : void
+    {
+        if (is_string($event->getTitle())) {
+            return;
+        }
+
+        $title = $event->getItem()->xpath()
+            ->firstOf('/libero:item/jats:article/jats:front/jats:article-meta/jats:title-group/jats:article-title');
```
Comment (giorgiosironi) [inline vendor-extra/JatsContentBundle/src/EventListener/TitleListener.php]: seems very specific - constant? constructor argument?

File: vendor-extra/JatsContentBundle/src/EventListener/ItemTagsListener.php [hunk 5]
```diff
@@ -20,20 +19,22 @@ public function __construct(ViewConverter $converter)
...
+    public function onCreatePageMain(CreateContentPagePartEvent $event) : void
     {
         $front = $event->getItem()->xpath()->firstOf('/libero:item/jats:article/jats:front');
 
         if (!$front instanceof Element) {
             return;
         }
 
-        $itemTags = $this->converter->convert($front, '@LiberoPatterns/item-tags.html.twig', $event->getContext());
+        $context = ['area' => 'primary'] + $event->getContext();
```
Comment (giorgiosironi) [inline vendor-extra/JatsContentBundle/src/EventListener/ItemTagsListener.php]: `primary` as a term will be part of a very popular interface, what are the alternatives for this value (when the `area` is not `primary`)? Is it standard teminology from Twig or another library?

Comment (thewilkybarkid) [reply vendor-extra/JatsContentBundle/src/EventListener/ItemTagsListener.php]: Refers to the area in the grid pattern: https://github.com/libero/pattern-library/blob/80b20661bb84aba1dbdc533bfcd5e13a3c272337/source/css/sass/patterns/templates/content-grid.scss#L69-L88

File: composer.json (hunk 3 above)
Comment (thewilkybarkid) [reply composer.json]: It's now there, but can't seem to get GitHub to produce a patch file that applies. Will just wait for the next release.

File: vendor-extra/JatsContentBundle/src/EventListener/TitleListener.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+namespace Libero\JatsContentBundle\EventListener;
+
+use FluentDOM\DOM\Element;
+use Libero\ContentPageBundle\Event\CreateContentPageEvent;
+use function is_string;
+
+final class TitleListener
+{
+    private const FRONT_PATH = '/libero:item/jats:article/jats:front';
+    private const TITLE_PATH = self::FRONT_PATH.'/jats:article-meta/jats:title-group/jats:article-title';
```
Comment (thewilkybarkid) [inline vendor-extra/JatsContentBundle/src/EventListener/TitleListener.php]: Had to split as the line is too long, and splitting over two lines doesn't see it indent how it probably should:

I'd expect

```php
    private const TITLE_PATH = '/libero:item/jats:article/jats:front/jats:article-meta/jats:title-group'
        .'/jats:article-title';
```

rather than

```php
    private const TITLE_PATH = '/libero:item/jats:article/jats:front/jats:article-meta/jats:title-group'
    .'/jats:article-title';
```

But probably better to avoid such long lines anyway?

File: vendor-extra/JatsContentBundle/src/EventListener/ItemTagsListener.php (hunk 5 above)
Comment (thewilkybarkid) [reply vendor-extra/JatsContentBundle/src/EventListener/ItemTagsListener.php]: Related: b66733c

File: vendor-extra/JatsContentBundle/src/EventListener/BodyListener.php [hunk 7]
```diff
@@ -5,39 +5,37 @@
...
 
         if (!$body instanceof Element) {
             return;
         }
 
-        $context = ['level' => ($event->getContext()['level'] ?? 1) + 1] + $event->getContext();
+        $context = [
+                'area' => CONTENT_GRID_PRIMARY,
+                'level' => ($event->getContext()['level'] ?? 1) + 1,
+            ] + $event->getContext();
```
Comment (thewilkybarkid) [inline vendor-extra/JatsContentBundle/src/EventListener/BodyListener.php]: IDE indentation rather than the coding standard. Looks a little odd.
